File: sample/src/main/java/com/marverenic/options/demo/MainActivity.java
package com.marverenic.options.demo;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.provider.Settings;
import android.support.v7.app.AppCompatActivity;

import com.marverenic.options.Option;
import com.marverenic.options.OptionFragment;
import com.marverenic.options.types.ActivityOption;
import com.marverenic.options.types.FragmentOption;
import com.marverenic.options.types.MultiSelectOption;
import com.marverenic.options.types.OptionHeader;
import com.marverenic.options.types.sharedpreferences.SharedPreferenceSwitchOption;
import com.marverenic.options.types.sharedpreferences.SharedPreferencesCheckOption;
import com.marverenic.options.types.sharedpreferences.dialog.SharedPreferencesIntDialogOption;
import com.marverenic.options.types.sharedpreferences.dialog.SharedPreferencesStringDialogOption;
import com.marverenic.options.types.sharedpreferences.dropdown.SharedPreferencesIntDropdownOption;
import com.marverenic.options.types.sharedpreferences.dropdown.SharedPreferencesStringDropdownOption;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private static final String SUB_SETTINGS_TAG = "subpanel";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        getSupportFragmentManager().beginTransaction()
                .replace(R.id.preference_container, new PreferenceFragment())
                .addToBackStack(null)
                .commit();
    }

    @Override
    public void onBackPressed() {
        if (!getSupportFragmentManager().popBackStackImmediate()) {
            super.onBackPressed();
        }
    }

    public static class PreferenceFragment extends OptionFragment {

        @Override
        public List<Option> createOptionList() {
            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());

            return Arrays.asList(
                    new OptionHeader("Toggle Options"),
                    new SharedPreferenceSwitchOption.Builder()
                            .setSharedPreferences(prefs)
                            .setKey("switch-1")
                            .setTitle("Switch 1")
                            .setEnabledDescription("Switch 1 is enabled")
                            .setDisabledDescription("Switch 1 is disabled")
                            .build(),
                    new SharedPreferencesCheckOption.Builder()
                            .setSharedPreferences(prefs)
                            .setKey("check-1")
                            .setTitle("CheckBox 1")
                            .setEnabledDescription("CheckBox 1 is enabled")
                            .setDisabledDescription("CheckBox 1 is disabled")
                            .setDefaultValue(true)
                            .build(),
                    new OptionHeader("Dropdown Options"),
                    new SharedPreferencesIntDropdownOption.Builder()
                            .setSharedPreferences(prefs)
                            .setKey("int-select-1")
                            .setTitle("Integer Dropdown")
                            .setDefaultValue(4)
                            .setValues(Arrays.asList(
                                    new MultiSelectOption.Selection<>("One", 1),
                                    new MultiSelectOption.Selection<>("Two", 2),
                                    new MultiSelectOption.Selection<>("Three", 3),
                                    new MultiSelectOption.Selection<>("Four", 4)
                            ))
                            .build(),
                    new SharedPreferencesStringDropdownOption.Builder()
                            .setSharedPreferences(prefs)
                            .setKey("string-select-1")
                            .setTitle("String Dropdown")
                            .setDefaultValue("medium")
                            .setValues(Arrays.asList(
                                    new MultiSelectOption.Selection<>("Never", "low"),
                                    new MultiSelectOption.Selection<>("Sometimes", "medium"),
                                    new MultiSelectOption.Selection<>("Always", "high")
                            ))
                            .build(),
                    new OptionHeader("Dialog Options"),
                    new SharedPreferencesIntDialogOption.Builder()
                            .setSharedPreferences(prefs)
                            .setKey("int-select-2")
                            .setTitle("Integer Dialog")
                            .setDefaultValue(0)
                            .setValues(Arrays.asList(
                                    new MultiSelectOption.Selection<>("None", 0),
                                    new MultiSelectOption.Selection<>("One", 1),
                                    new MultiSelectOption.Selection<>("A Couple", 2),
                                    new MultiSelectOption.Selection<>("Several", 3),
                                    new MultiSelectOption.Selection<>("Many", 4)
                            ))
                            .build(),
                    new SharedPreferencesStringDialogOption.Builder()
                            .setSharedPreferences(prefs)
                            .setKey("string-select-2")
                            .setTitle("String Dialog")
                            .setDefaultValue("orange")
                            .setValues(Arrays.asList(
                                    new MultiSelectOption.Selection<>("Red", "red"),
                                    new MultiSelectOption.Selection<>("Orange", "orange"),
                                    new MultiSelectOption.Selection<>("Yellow", "yellow"),
                                    new MultiSelectOption.Selection<>("Green", "green")
                            ))
                            .build(),
                    new OptionHeader("Link Options"),
                    new ActivityOption("Device Settings", "Launch settings activity",
                            new Intent(Settings.ACTION_SETTINGS)),
                    new FragmentOption("More Settings", "Submenu",
                            new SubPreferenceFragment(), SUB_SETTINGS_TAG)
            );
        }
    }

    public static class SubPreferenceFragment extends OptionFragment {

        @Override
        public List<Option> createOptionList() {
            return Collections.emptyList();
        }
    }
}

File: sample-kotlin/src/main/java/com/marverenic/options/demo/MainActivity.kt
package com.marverenic.options.demo

import android.content.Intent
import android.os.Bundle
import android.preference.PreferenceManager
import android.provider.Settings
import android.support.v7.app.AppCompatActivity
import com.marverenic.options.Option
import com.marverenic.options.OptionFragment
import com.marverenic.options.demo.kotlin.R
import com.marverenic.options.kotlin.optionsOf
import com.marverenic.options.kotlin.selectionsOf

private const val SUB_SETTINGS_TAG = "subpanel"

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        supportFragmentManager.beginTransaction()
                .replace(R.id.preference_container, PreferenceFragment())
                .commit()
    }

    class PreferenceFragment : OptionFragment() {

        override fun createOptionList(): List<Option> {
            val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
            return optionsOf(sharedPreferences) {
                header(title = "Toggle Options")
                switchOption(key = "switch-1") {
                    title = "Switch 1"
                    enabledDescription = "Switch 1 is enabled"
                    disabledDescription = "Switch 1 is disabled"
                }
                checkOption(key = "check-1") {
                    title = "CheckBox 1"
                    enabledDescription = "CheckBox 1 is enabled"
                    disabledDescription = "CheckBox 1 is disabled"
                }


                header(title = "Dropdown options")
                intDropdownOption(key = "int-select-1") {
                    title = "Integer Dropdown"
                    defaultValue = 4
                    values = selectionsOf(
                            "One" to 1,
                            "Two" to 2,
                            "Three" to 3,
                            "Four" to 4
                    )
                }
                stringDropdownOption(key = "string-select-1") {
                    title = "String Dropdown"
                    defaultValue = "medium"
                    values = selectionsOf(
                            "Never" to "low",
                            "Sometimes" to "medium",
                            "Always" to "high"
                    )
                }


                header(title = "Dialog options")
                intDialogOption(key = "int-select-2") {
                    title = "Integer Dialog"
                    defaultValue = 0
                    values = selectionsOf(
                            "None" to 0,
                            "One" to 1,
                            "A Couple" to 2,
                            "Several" to 3,
                            "Mnay" to 4
                    )
                }
                stringDialogOption(key = "string-select-2") {
                    title = "String Dialog"
                    defaultValue = "orange"
                    values = selectionsOf(
                            "Red" to "red",
                            "Orange" to "orange",
                            "Yellow" to "yellow",
                            "Green" to "green"
                    )
                }

                header(title = "Link Options")
                activityOption(
                        title = "Device settings",
                        description = "Open settings activity",
                        intent = Intent(Settings.ACTION_SETTINGS)
                )
                fragmentOption(
                        title = "More Settings",
                        description = "Submenu",
                        fragment = SubPreferenceFragment(),
                        tag = SUB_SETTINGS_TAG
                )
            }
        }
    }

    class SubPreferenceFragment : OptionFragment() {

        override fun createOptionList(): List<Option> {
            return emptyList<Option>()
        }


    }
}
